# ETL Pipeline Preparation
Follow the instructions below to help you create your ETL pipeline.
### 1. Import libraries and load datasets.
- Import Python libraries
- Load `messages.csv` into a dataframe and inspect the first few lines.
- Load `categories.csv` into a dataframe and inspect the first few lines.

In [1]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

from sqlalchemy import create_engine
pd.options.display.max_rows = 999
pd.options.display.max_columns = 200

In [2]:
# load messages dataset
messages = pd.read_csv('messages.csv')
messages.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  
0  Un front froid se retrouve sur Cuba ce matin. ...  direct  
1                 Cyclone nan fini osinon li pa fini  direct  
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct  
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
4  facade ouest d Haiti et le reste du pays aujou...  direct

In [3]:
# Check missing values, data type, etc.
messages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26248 entries, 0 to 26247
Data columns (total 4 columns):
id          26248 non-null int64
message     26248 non-null object
original    10184 non-null object
genre       26248 non-null object
dtypes: int64(1), object(3)
memory usage: 820.3+ KB


In [4]:
# Check duplicated data
messages.id.nunique()

26180

In [5]:
# load categories dataset
categories = pd.read_csv('categories.csv')
categories.head()

id                                         categories
0   2  related-1;request-0;offer-0;aid_related-0;medi...
1   7  related-1;request-0;offer-0;aid_related-1;medi...
2   8  related-1;request-0;offer-0;aid_related-0;medi...
3   9  related-1;request-1;offer-0;aid_related-1;medi...
4  12  related-1;request-0;offer-0;aid_related-0;medi...

In [6]:
# Check missing values, data type, etc.
categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26248 entries, 0 to 26247
Data columns (total 2 columns):
id            26248 non-null int64
categories    26248 non-null object
dtypes: int64(1), object(1)
memory usage: 410.2+ KB


In [7]:
# Check duplicated data
categories.id.nunique()

26180

The messages dataset has 26248 rows, which contains 26180 unique ids. The original column contains more than 50% missing values.  
The categories dataset has 26248 rows, which contains 26180 uniques ids. The categories column contains various categories in each row.  

### 2. Merge datasets.
- Merge the messages and categories datasets using the common id
- Assign this combined dataset to `df`, which will be cleaned in the following steps

In [8]:
# merge datasets
df = messages.merge(categories, on='id', how='left')
df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  \
0  Un front froid se retrouve sur Cuba ce matin. ...  direct   
1                 Cyclone nan fini osinon li pa fini  direct   
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct   
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct   
4  facade ouest d Haiti et le reste du pays aujou...  direct   

                                          categories  
0  related-1;request-0;offer-0;aid_related-0;medi...  
1  related-1;request-0;offer-0;aid_related-1;medi...  
2  related-1;request-0;offer-0;aid_related-0;medi...  
3  related-1;request-1;offer-0;aid_related-1;medi...  
4  related-1;request-0;offer-0;aid_related-0;medi...

In [9]:
# Check general info
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26386 entries, 0 to 26385
Data columns (total 5 columns):
id            26386 non-null int64
message       26386 non-null object
original      10246 non-null object
genre         26386 non-null object
categories    26386 non-null object
dtypes: int64(1), object(4)
memory usage: 1.2+ MB


There are 26386 rows in the combined dataset. The datasets before combination have 26248 rows. It is possible there are duplicated data.  

### 3. Split `categories` into separate category columns.
- Split the values in the `categories` column on the `;` character so that each value becomes a separate column. You'll find [this method](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.Series.str.split.html) very helpful! Make sure to set `expand=True`.
- Use the first row of categories dataframe to create column names for the categories data.
- Rename columns of `categories` with new column names.

In [10]:
# create a dataframe of the 36 individual category columns
categories = df['categories'].str.split(';',expand=True)
categories.head()

0          1        2              3               4   \
0  related-1  request-0  offer-0  aid_related-0  medical_help-0   
1  related-1  request-0  offer-0  aid_related-1  medical_help-0   
2  related-1  request-0  offer-0  aid_related-0  medical_help-0   
3  related-1  request-1  offer-0  aid_related-1  medical_help-0   
4  related-1  request-0  offer-0  aid_related-0  medical_help-0   

                   5                    6           7           8   \
0  medical_products-0  search_and_rescue-0  security-0  military-0   
1  medical_products-0  search_and_rescue-0  security-0  military-0   
2  medical_products-0  search_and_rescue-0  security-0  military-0   
3  medical_products-1  search_and_rescue-0  security-0  military-0   
4  medical_products-0  search_and_rescue-0  security-0  military-0   

              9        10      11         12          13       14  \
0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
1  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
2  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
3  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
4  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   

                 15          16       17           18  \
0  missing_people-0  refugees-0  death-0  other_aid-0   
1  missing_people-0  refugees-0  death-0  other_aid-1   
2  missing_people-0  refugees-0  death-0  other_aid-0   
3  missing_people-0  refugees-0  death-0  other_aid-1   
4  missing_people-0  refugees-0  death-0  other_aid-0   

                         19           20           21             22       23  \
0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   
1  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   
2  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   
3  infrastructure_related-1  transport-0  buildings-1  electricity-0  tools-0   
4  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   

            24       25             26                      27  \
0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   
1  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   
2  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   
3  hospitals-1  shops-0  aid_centers-0  other_infrastructure-0   
4  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   

                  28        29       30      31            32      33  \
0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
1  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
2  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
3  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
4  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

                34               35  
0  other_weather-0  direct_report-0  
1  other_weather-0  direct_report-0  
2  other_weather-0  direct_report-0  
3  other_weather-0  direct_report-0  
4  other_weather-0  direct_report-0

In [11]:
# select the first row of the categories dataframe
row = categories.iloc[0]

# use this row to extract a list of new column names for categories.
# one way is to apply a lambda function that takes everything 
# up to the second to last character of each string with slicing
category_colnames = categories.iloc[0].apply(lambda x: x[:-2]).values
print(category_colnames)

['related' 'request' 'offer' 'aid_related' 'medical_help'
 'medical_products' 'search_and_rescue' 'security' 'military' 'child_alone'
 'water' 'food' 'shelter' 'clothing' 'money' 'missing_people' 'refugees'
 'death' 'other_aid' 'infrastructure_related' 'transport' 'buildings'
 'electricity' 'tools' 'hospitals' 'shops' 'aid_centers'
 'other_infrastructure' 'weather_related' 'floods' 'storm' 'fire'
 'earthquake' 'cold' 'other_weather' 'direct_report']


In [12]:
# rename the columns of `categories`
categories.columns = category_colnames
categories.head()

related    request    offer    aid_related    medical_help  \
0  related-1  request-0  offer-0  aid_related-0  medical_help-0   
1  related-1  request-0  offer-0  aid_related-1  medical_help-0   
2  related-1  request-0  offer-0  aid_related-0  medical_help-0   
3  related-1  request-1  offer-0  aid_related-1  medical_help-0   
4  related-1  request-0  offer-0  aid_related-0  medical_help-0   

     medical_products    search_and_rescue    security    military  \
0  medical_products-0  search_and_rescue-0  security-0  military-0   
1  medical_products-0  search_and_rescue-0  security-0  military-0   
2  medical_products-0  search_and_rescue-0  security-0  military-0   
3  medical_products-1  search_and_rescue-0  security-0  military-0   
4  medical_products-0  search_and_rescue-0  security-0  military-0   

     child_alone    water    food    shelter    clothing    money  \
0  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
1  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
2  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
3  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   
4  child_alone-0  water-0  food-0  shelter-0  clothing-0  money-0   

     missing_people    refugees    death    other_aid  \
0  missing_people-0  refugees-0  death-0  other_aid-0   
1  missing_people-0  refugees-0  death-0  other_aid-1   
2  missing_people-0  refugees-0  death-0  other_aid-0   
3  missing_people-0  refugees-0  death-0  other_aid-1   
4  missing_people-0  refugees-0  death-0  other_aid-0   

     infrastructure_related    transport    buildings    electricity    tools  \
0  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   
1  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   
2  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   
3  infrastructure_related-1  transport-0  buildings-1  electricity-0  tools-0   
4  infrastructure_related-0  transport-0  buildings-0  electricity-0  tools-0   

     hospitals    shops    aid_centers    other_infrastructure  \
0  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   
1  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   
2  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   
3  hospitals-1  shops-0  aid_centers-0  other_infrastructure-0   
4  hospitals-0  shops-0  aid_centers-0  other_infrastructure-0   

     weather_related    floods    storm    fire    earthquake    cold  \
0  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
1  weather_related-1  floods-0  storm-1  fire-0  earthquake-0  cold-0   
2  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
3  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   
4  weather_related-0  floods-0  storm-0  fire-0  earthquake-0  cold-0   

     other_weather    direct_report  
0  other_weather-0  direct_report-0  
1  other_weather-0  direct_report-0  
2  other_weather-0  direct_report-0  
3  other_weather-0  direct_report-0  
4  other_weather-0  direct_report-0

### 4. Convert category values to just numbers 0 or 1.
- Iterate through the category columns in df to keep only the last character of each string (the 1 or 0). For example, `related-0` becomes `0`, `related-1` becomes `1`. Convert the string to a numeric value.
- You can perform [normal string actions on Pandas Series](https://pandas.pydata.org/pandas-docs/stable/text.html#indexing-with-str), like indexing, by including `.str` after the Series. You may need to first convert the Series to be of type string, which you can do with `astype(str)`.

In [13]:
for column in categories:
    # set each value to be the last character of the string
    categories[column] = categories[column].apply(lambda x:int(x[-1]))
categories.head()

related  request  offer  aid_related  medical_help  medical_products  \
0        1        0      0            0             0                 0   
1        1        0      0            1             0                 0   
2        1        0      0            0             0                 0   
3        1        1      0            1             0                 1   
4        1        0      0            0             0                 0   

   search_and_rescue  security  military  child_alone  water  food  shelter  \
0                  0         0         0            0      0     0        0   
1                  0         0         0            0      0     0        0   
2                  0         0         0            0      0     0        0   
3                  0         0         0            0      0     0        0   
4                  0         0         0            0      0     0        0   

   clothing  money  missing_people  refugees  death  other_aid  \
0         0      0               0         0      0          0   
1         0      0               0         0      0          1   
2         0      0               0         0      0          0   
3         0      0               0         0      0          1   
4         0      0               0         0      0          0   

   infrastructure_related  transport  buildings  electricity  tools  \
0                       0          0          0            0      0   
1                       0          0          0            0      0   
2                       0          0          0            0      0   
3                       1          0          1            0      0   
4                       0          0          0            0      0   

   hospitals  shops  aid_centers  other_infrastructure  weather_related  \
0          0      0            0                     0                0   
1          0      0            0                     0                1   
2          0      0            0                     0                0   
3          1      0            0                     0                0   
4          0      0            0                     0                0   

   floods  storm  fire  earthquake  cold  other_weather  direct_report  
0       0      0     0           0     0              0              0  
1       0      1     0           0     0              0              0  
2       0      0     0           0     0              0              0  
3       0      0     0           0     0              0              0  
4       0      0     0           0     0              0              0

In [14]:
# check whether all the category columns are binary
for column in categories:
    print(column,categories[column].unique())

related [1 0 2]
request [0 1]
offer [0 1]
aid_related [0 1]
medical_help [0 1]
medical_products [0 1]
search_and_rescue [0 1]
security [0 1]
military [0 1]
child_alone [0]
water [0 1]
food [0 1]
shelter [0 1]
clothing [0 1]
money [0 1]
missing_people [0 1]
refugees [0 1]
death [0 1]
other_aid [0 1]
infrastructure_related [0 1]
transport [0 1]
buildings [0 1]
electricity [0 1]
tools [0 1]
hospitals [0 1]
shops [0 1]
aid_centers [0 1]
other_infrastructure [0 1]
weather_related [0 1]
floods [0 1]
storm [0 1]
fire [0 1]
earthquake [0 1]
cold [0 1]
other_weather [0 1]
direct_report [0 1]


In [15]:
# Check what each value is in the rows with related column equals to 2
categories[categories['related']==2]

related  request  offer  aid_related  medical_help  medical_products  \
117          2        0      0            0             0                 0   
221          2        0      0            0             0                 0   
307          2        0      0            0             0                 0   
462          2        0      0            0             0                 0   
578          2        0      0            0             0                 0   
657          2        0      0            0             0                 0   
658          2        0      0            0             0                 0   
659          2        0      0            0             0                 0   
660          2        0      0            0             0                 0   
889          2        0      0            0             0                 0   
903          2        0      0            0             0                 0   
931          2        0      0            0             0                 0   
937          2        0      0            0             0                 0   
939          2        0      0            0             0                 0   
1234         2        0      0            0             0                 0   
1256         2        0      0            0             0                 0   
1317         2        0      0            0             0                 0   
1409         2        0      0            0             0                 0   
1506         2        0      0            0             0                 0   
1694         2        0      0            0             0                 0   
1788         2        0      0            0             0                 0   
2354         2        0      0            0             0                 0   
2479         2        0      0            0             0                 0   
2544         2        0      0            0             0                 0   
2560         2        0      0            0             0                 0   
3033         2        0      0            0             0                 0   
3138         2        0      0            0             0                 0   
3374         2        0      0            0             0                 0   
3630         2        0      0            0             0                 0   
4609         2        0      0            0             0                 0   
4651         2        0      0            0             0                 0   
4652         2        0      0            0             0                 0   
4653         2        0      0            0             0                 0   
4654         2        0      0            0             0                 0   
4746         2        0      0            0             0                 0   
4805         2        0      0            0             0                 0   
5107         2        0      0            0             0                 0   
5587         2        0      0            0             0                 0   
5588         2        0      0            0             0                 0   
5589         2        0      0            0             0                 0   
5590         2        0      0            0             0                 0   
5921         2        0      0            0             0                 0   
6020         2        0      0            0             0                 0   
6119         2        0      0            0             0                 0   
6304         2        0      0            0             0                 0   
6386         2        0      0            0             0                 0   
6461         2        0      0            0             0                 0   
6518         2        0      0            0             0                 0   
6522         2        0      0            0             0                 0   
7189         2        0      0            0             0

Seems all the values are 0. Hence I will change all the 2 to 0.

In [16]:
# change all the values from 2 to 1 in the related column
categories.loc[categories['related']==2,'related']=0

In [17]:
# check whether the change works
categories['related'].unique()

array([1, 0])

### 5. Replace `categories` column in `df` with new category columns.
- Drop the categories column from the df dataframe since it is no longer needed.
- Concatenate df and categories data frames.

In [18]:
# drop the original categories column from `df`
df=df.drop('categories',axis=1)

df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  
0  Un front froid se retrouve sur Cuba ce matin. ...  direct  
1                 Cyclone nan fini osinon li pa fini  direct  
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct  
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct  
4  facade ouest d Haiti et le reste du pays aujou...  direct

In [19]:
# concatenate the original dataframe with the new `categories` dataframe
df = pd.concat([df,categories],axis=1)
df.head()

id                                            message  \
0   2  Weather update - a cold front from Cuba that c...   
1   7            Is the Hurricane over or is it not over   
2   8                    Looking for someone but no name   
3   9  UN reports Leogane 80-90 destroyed. Only Hospi...   
4  12  says: west side of Haiti, rest of the country ...   

                                            original   genre  related  \
0  Un front froid se retrouve sur Cuba ce matin. ...  direct        1   
1                 Cyclone nan fini osinon li pa fini  direct        1   
2  Patnm, di Maryani relem pou li banm nouvel li ...  direct        1   
3  UN reports Leogane 80-90 destroyed. Only Hospi...  direct        1   
4  facade ouest d Haiti et le reste du pays aujou...  direct        1   

   request  offer  aid_related  medical_help  medical_products  \
0        0      0            0             0                 0   
1        0      0            1             0                 0   
2        0      0            0             0                 0   
3        1      0            1             0                 1   
4        0      0            0             0                 0   

   search_and_rescue  security  military  child_alone  water  food  shelter  \
0                  0         0         0            0      0     0        0   
1                  0         0         0            0      0     0        0   
2                  0         0         0            0      0     0        0   
3                  0         0         0            0      0     0        0   
4                  0         0         0            0      0     0        0   

   clothing  money  missing_people  refugees  death  other_aid  \
0         0      0               0         0      0          0   
1         0      0               0         0      0          1   
2         0      0               0         0      0          0   
3         0      0               0         0      0          1   
4         0      0               0         0      0          0   

   infrastructure_related  transport  buildings  electricity  tools  \
0                       0          0          0            0      0   
1                       0          0          0            0      0   
2                       0          0          0            0      0   
3                       1          0          1            0      0   
4                       0          0          0            0      0   

   hospitals  shops  aid_centers  other_infrastructure  weather_related  \
0          0      0            0                     0                0   
1          0      0            0                     0                1   
2          0      0            0                     0                0   
3          1      0            0                     0                0   
4          0      0            0                     0                0   

   floods  storm  fire  earthquake  cold  other_weather  direct_report  
0       0      0     0           0     0              0              0  
1       0      1     0           0     0              0              0  
2       0      0     0           0     0              0              0  
3       0      0     0           0     0              0              0  
4       0      0     0           0     0              0              0

In [20]:
# Check general info
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26386 entries, 0 to 26385
Data columns (total 40 columns):
id                        26386 non-null int64
message                   26386 non-null object
original                  10246 non-null object
genre                     26386 non-null object
related                   26386 non-null int64
request                   26386 non-null int64
offer                     26386 non-null int64
aid_related               26386 non-null int64
medical_help              26386 non-null int64
medical_products          26386 non-null int64
search_and_rescue         26386 non-null int64
security                  26386 non-null int64
military                  26386 non-null int64
child_alone               26386 non-null int64
water                     26386 non-null int64
food                      26386 non-null int64
shelter                   26386 non-null int64
clothing                  26386 non-null int64
money                     26386 non-null i

### 6. Remove duplicates.
- Check how many duplicates are in this dataset.
- Drop the duplicates.
- Confirm duplicates were removed.

In [21]:
# check number of duplicates
df.loc[df.duplicated()]

id                                            message  \
164      202  ?? port au prince ?? and food. they need gover...   
165      202  ?? port au prince ?? and food. they need gover...   
658      804  elle est vraiment malade et a besoin d'aide. u...   
659      804  elle est vraiment malade et a besoin d'aide. u...   
660      804  elle est vraiment malade et a besoin d'aide. u...   
715      862  What is the address of the radio station? I as...   
716      862  What is the address of the radio station? I as...   
1415    1652  please we need water, food and tents, we have ...   
1416    1652  please we need water, food and tents, we have ...   
2139    2446  How much money did TV Latino American collect ...   
2140    2446  How much money did TV Latino American collect ...   
2141    2446  How much money did TV Latino American collect ...   
2803    3217  Sir or madam, I'm hungry, I cannot stand becau...   
2804    3217  Sir or madam, I'm hungry, I cannot stand becau...   
2805    3217  Sir or madam, I'm hungry, I cannot stand becau...   
2838    3250  I thought you gave my number and this was seri...   
2839    3250  I thought you gave my number and this was seri...   
3389    3882  I am in Carefour Feuilles, Dkayet neighborhood...   
3390    3882  I am in Carefour Feuilles, Dkayet neighborhood...   
3868    4412  Good evening staff of responsibles, please giv...   
3869    4412  Good evening staff of responsibles, please giv...   
3870    4412  Good evening staff of responsibles, please giv...   
4360    4956  What we saw yesterday is a sign, you need to k...   
4361    4956  What we saw yesterday is a sign, you need to k...   
4543    5153  there are people in Delmas 32 who form an asso...   
4544    5153  there are people in Delmas 32 who form an asso...   
4652    5265  (Delmas 33 Charboniere infomatyon s'il vous pl...   
4653    5265  (Delmas 33 Charboniere infomatyon s'il vous pl...   
4654    5265  (Delmas 33 Charboniere infomatyon s'il vous pl...   
4917    5573  If I would like to find a job in cepoz, what s...   
4918    5573  If I would like to find a job in cepoz, what s...   
4919    5573  If I would like to find a job in cepoz, what s...   
4977    5643  Good evening. we are writing to ask you for a ...   
4978    5643  Good evening. we are writing to ask you for a ...   
4979    5643  Good evening. we are writing to ask you for a ...   
5101    5776  I would like some more informations about trav...   
5102    5776  I would like some more informations about trav...   
5588    6327  jel2  Acte 5: 29 2chr7 Lev11 Ecl9 Ecl4 jos5 ch...   
5589    6327  jel2  Acte 5: 29 2chr7 Lev11 Ecl9 Ecl4 jos5 ch...   
5590    6327  jel2  Acte 5: 29 2chr7 Lev11 Ecl9 Ecl4 jos5 ch...   
5640    6393    I'd like to sign up in the 4636 program please    
5641    6393    I'd like to sign up in the 4636 program please    
5732    6492           I'd like to know an email to find a job    
5733    6492           I'd like to know an email to find a job    
5757    6515  I would like to find some informations about t...   
5758    6515  I would like to find some informations about t...   
5810    6573    I am a victim. I need food, water and a radio.    
5811    6573    I am a victim. I need food, water and a radio.    
5812    6573    I am a victim. I need food, water and a radio.    
5919    6687  Tell the President to give the nation over to ...   
5920    6687  Tell the President to give the nation over to ...   
6887    7747  where we can paticipate in the law reorganizat...   
6888    7747  where we can paticipate in the law reorganizat...   
7027    7892  Good morning!How are you doing my brothers whe...   
7028    7892  Good morning!How are you doing my brothers whe...   
7029    7892  Good morning!How are you doing my brothers whe...   
7084    7945  What can i do to protect my self if the earthq...   
7085    7945  What can i do to protect my self if the earthq...   
7313    8190  here,i'm i don't find medicin just resting 

In [22]:
# drop duplicates
df.drop_duplicates(inplace=True)

In [23]:
# check number of duplicates
df.loc[df.duplicated()]

Empty DataFrame
Columns: [id, message, original, genre, related, request, offer, aid_related, medical_help, medical_products, search_and_rescue, security, military, child_alone, water, food, shelter, clothing, money, missing_people, refugees, death, other_aid, infrastructure_related, transport, buildings, electricity, tools, hospitals, shops, aid_centers, other_infrastructure, weather_related, floods, storm, fire, earthquake, cold, other_weather, direct_report]
Index: []

In [24]:
# Drop the original column since it has more than 50% missing values
df.drop('original',axis=1,inplace=True)

In [25]:
# Check general info
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26216 entries, 0 to 26385
Data columns (total 39 columns):
id                        26216 non-null int64
message                   26216 non-null object
genre                     26216 non-null object
related                   26216 non-null int64
request                   26216 non-null int64
offer                     26216 non-null int64
aid_related               26216 non-null int64
medical_help              26216 non-null int64
medical_products          26216 non-null int64
search_and_rescue         26216 non-null int64
security                  26216 non-null int64
military                  26216 non-null int64
child_alone               26216 non-null int64
water                     26216 non-null int64
food                      26216 non-null int64
shelter                   26216 non-null int64
clothing                  26216 non-null int64
money                     26216 non-null int64
missing_people            26216 non-null in

In [26]:
# Check whether there are missing values
df.isnull().sum().sum()

0

### 7. Save the clean dataset into an sqlite database.
You can do this with pandas [`to_sql` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) combined with the SQLAlchemy library. Remember to import SQLAlchemy's `create_engine` in the first cell of this notebook to use it below.

In [27]:
engine = create_engine('sqlite:///DisasterResponse.db')
df.to_sql('DisasterRecord', con=engine, if_exists = 'replace', index=False)

### 8. Use this notebook to complete `etl_pipeline.py`
Use the template file attached in the Resources folder to write a script that runs the steps above to create a database based on new datasets specified by the user. Alternatively, you can complete `etl_pipeline.py` in the classroom on the `Project Workspace IDE` coming later.